src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -1894,5 +2044,681 @@ struct S
+        }
+
+        [Fact]
+        public void Subsumption_12()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        _ = a switch
+        {
+            { Length: not 1 }  => 0,
+            [<0, ..] => 0,
+            [..[>= 0]] or [..null] => 1,
+            [_] => 2, // unreachable
+        };
+        _ = a switch // exhaustive
+        {
+            { Length: not 1 }  => 0,
+            [<0, ..] => 0,
+            [..[>= 0]] => 1,
+            [_] => 2,
+        };
+    }
+}" + TestSources.GetSubArray;
+            var comp = CreateCompilationWithIndexAndRange(src);
+            comp.VerifyEmitDiagnostics(
+                // (11,13): error CS8510: The pattern is unreachable. It has already been handled by a previous arm of the switch expression or it is impossible to match.
+                //             [_] => 2, // unreachable
+                Diagnostic(ErrorCode.ERR_SwitchArmSubsumed, "[_]").WithLocation(11, 13));
+        }
+
+        [Fact]
+        public void Subsumption_13()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        _ = a switch
+        {
+            [.., >0] => 1,
+            [<0, ..] => 2,
+            [0, ..] => 3,
+            { Length: not 1 } => 4,
+            [var unreachable] => 5,
+        };
+    }
+}" + TestSources.GetSubArray;
+            var comp = CreateCompilationWithIndexAndRange(src);
+            comp.VerifyEmitDiagnostics(
+                // (12,13): error CS8510: The pattern is unreachable. It has already been handled by a previous arm of the switch expression or it is impossible to match.
+                //             [var unreachable] => 5,
+                Diagnostic(ErrorCode.ERR_SwitchArmSubsumed, "[var unreachable]").WithLocation(12, 13));
+        }
+
+        [Fact]
+        public void Subsumption_14()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        _ = a switch
+        {
+            [.., >=0, >0] => 0,
+            [.., <0, >=0] => 0,
+            [<=0, <0, ..] => 1,
+            [>0, <=0, ..] => 1,
+            [0, 0] => 1,
+            { Length: not 2 }  => 2,
+            [var unreachable, var unreachable2] => 3,
+        };
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src);
+            comp.VerifyEmitDiagnostics(
+                // (14,13): error CS8510: The pattern is unreachable. It has already been handled by a previous arm of the switch expression or it is impossible to match.
+                //             [var unreachable, var unreachable2] => 3,
+                Diagnostic(ErrorCode.ERR_SwitchArmSubsumed, "[var unreachable, var unreachable2]").WithLocation(14, 13));
+        }
+
+        [Fact]
+        public void Subsumption_15()
+        {
+            // testing the scenario where we have multiple preconditions for indexers to relate.
```

Comment:
Below is the DAG for this test (see below).
Following the main path (state 0..8), I'm surprised to see that we test two values for `42` (at states 4 and 8) in one path. I would have expected all paths in the DAG to involve at most one such test.

<details><summary>Dump</summary>

```
            case [.., [.., 42]]:
            case [[42]]:

State 0
    1. [case [.., [.., 42]]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueOne(1: Int32)), t2=DagIndexerEvaluation(t0[-1]), ?DagNonNullTest(t2), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagNonNullTest(t0), t1=DagPropertyEvaluation(t0.Length), ?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagNonNullTest(t0)
    TrueBranch: 1
    FalseBranch: 27
State 1
    1. [case [.., [.., 42]]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueOne(1: Int32)), t2=DagIndexerEvaluation(t0[-1]), ?DagNonNullTest(t2), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t1=DagPropertyEvaluation(t0.Length)
    TrueBranch: 2
State 2
    1. [case [.., [.., 42]]:] AND(?DagRelationalTest(t1 >= ConstantValueOne(1: Int32)), t2=DagIndexerEvaluation(t0[-1]), ?DagNonNullTest(t2), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagRelationalTest(t1 >= ConstantValueOne(1: Int32))
    TrueBranch: 3
    FalseBranch: 27
State 3 REMAINING t1:[1..2147483647]
    1. [case [.., [.., 42]]:] AND(t2=DagIndexerEvaluation(t0[-1]), ?DagNonNullTest(t2), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t2=DagIndexerEvaluation(t0[-1])
    TrueBranch: 4
State 4 REMAINING t1:[1..2147483647]
    1. [case [.., [.., 42]]:] AND(?DagNonNullTest(t2), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagNonNullTest(t2)
    TrueBranch: 5
    FalseBranch: 24
State 5 REMAINING t1:[1..2147483647]
    1. [case [.., [.., 42]]:] AND(t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), OR(AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t3=DagPropertyEvaluation(t2.Length)
    TrueBranch: 6
State 6 REMAINING t1:[1..2147483647]
    1. [case [.., [.., 42]]:] AND(?DagRelationalTest(t3 >= ConstantValueOne(1: Int32)), t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), OR(AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagRelationalTest(t3 >= ConstantValueOne(1: Int32))
    TrueBranch: 7
    FalseBranch: 19
State 7 REMAINING t1:[1..2147483647] t3:[1..2147483647]
    1. [case [.., [.., 42]]:] AND(t4=DagIndexerEvaluation(t2[-1]), ?DagValueTest(t4 == ConstantValueI32(42: Int32)))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), OR(AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), OR(Not (?DagValueTest(t1 == ConstantValueOne(1: Int32))), DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t4=DagIndexerEvaluation(t2[-1])
    TrueBranch: 8
State 8 REMAINING t1:[1..2147483647] t3:[1..2147483647]
    1. [case [.., [.., 42]]:] ?DagValueTest(t4 == ConstantValueI32(42: Int32))
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), OR(AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), OR(Not (?DagValueTest(t1 == ConstantValueOne(1: Int32))), DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagValueTest(t4 == ConstantValueI32(42: Int32))
    TrueBranch: 9
    FalseBranch: 10
*State 9 REMAINING t1:[1..2147483647] t4:[42..42] t3:[1..2147483647]
    1. [case [.., [.., 42]]:] MATCH
State 10 REMAINING t1:[1..2147483647] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), OR(AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), OR(Not (?DagValueTest(t1 == ConstantValueOne(1: Int32))), DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), Not (AND(?DagValueTest(t3 == ConstantValueOne(1: Int32)), ?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagValueTest(t1 == ConstantValueOne(1: Int32))
    TrueBranch: 11
    FalseBranch: 27
State 11 REMAINING t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(t5=DagIndexerEvaluation(t0[0]), OR(DagAssignmentEvaluation(t5=t2), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), DagAssignmentEvaluation(t6=t3), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), Not (AND(?DagValueTest(t3 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t5=DagIndexerEvaluation(t0[0])
    TrueBranch: 12
State 12 REMAINING t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(OR(DagAssignmentEvaluation(t5=t2), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), DagAssignmentEvaluation(t6=t3), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), Not (AND(?DagValueTest(t3 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t5=t2)
    TrueBranch: 13
State 13 REMAINING t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(t6=DagPropertyEvaluation(t5.Length), DagAssignmentEvaluation(t6=t3), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), Not (AND(?DagValueTest(t3 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t6=DagPropertyEvaluation(t5.Length)
    TrueBranch: 14
State 14 REMAINING t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(DagAssignmentEvaluation(t6=t3), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), Not (AND(?DagValueTest(t3 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t6=t3)
    TrueBranch: 15
State 15 REMAINING t6:[1..2147483647] t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), Not (AND(?DagValueTest(t3 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagValueTest(t6 == ConstantValueOne(1: Int32))
    TrueBranch: 16
    FalseBranch: 27
State 16 REMAINING t6:[1..1] t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(t7=DagIndexerEvaluation(t5[0]), Not (AND(OR(DagAssignmentEvaluation(t3=t6), ?DagValueTest(t3 == ConstantValueOne(1: Int32))), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t7=DagIndexerEvaluation(t5[0])
    TrueBranch: 17
State 17 REMAINING t6:[1..1] t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..2147483647]
    2. [case [[42]]:] AND(Not (AND(OR(DagAssignmentEvaluation(t3=t6), ?DagValueTest(t3 == ConstantValueOne(1: Int32))), DagAssignmentEvaluation(t7=t4))), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t3=t6)
    TrueBranch: 18
State 18 REMAINING t6:[1..1] t1:[1..1] t4:[-2147483648..41],[43..2147483647] t3:[1..1]
    2. [case [[42]]:] AND(Not (DagAssignmentEvaluation(t7=t4)), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t7=t4)
    TrueBranch: 27
State 19 REMAINING t1:[1..2147483647] t3:[0..0]
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), OR(AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), Not (AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t6=t3))), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagValueTest(t1 == ConstantValueOne(1: Int32))
    TrueBranch: 20
    FalseBranch: 27
State 20 REMAINING t1:[1..1] t3:[0..0]
    2. [case [[42]]:] AND(t5=DagIndexerEvaluation(t0[0]), OR(DagAssignmentEvaluation(t5=t2), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), Not (DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t5=DagIndexerEvaluation(t0[0])
    TrueBranch: 21
State 21 REMAINING t1:[1..1] t3:[0..0]
    2. [case [[42]]:] AND(OR(DagAssignmentEvaluation(t5=t2), ?DagNonNullTest(t5)), t6=DagPropertyEvaluation(t5.Length), Not (DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t5=t2)
    TrueBranch: 22
State 22 REMAINING t1:[1..1] t3:[0..0]
    2. [case [[42]]:] AND(t6=DagPropertyEvaluation(t5.Length), Not (DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t6=DagPropertyEvaluation(t5.Length)
    TrueBranch: 23
State 23 REMAINING t1:[1..1] t3:[0..0]
    2. [case [[42]]:] AND(Not (DagAssignmentEvaluation(t6=t3)), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t6=t3)
    TrueBranch: 27
State 24 REMAINING t1:[1..2147483647]
    2. [case [[42]]:] AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), t5=DagIndexerEvaluation(t0[0]), Not (AND(?DagValueTest(t1 == ConstantValueOne(1: Int32)), DagAssignmentEvaluation(t5=t2))), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: ?DagValueTest(t1 == ConstantValueOne(1: Int32))
    TrueBranch: 25
    FalseBranch: 27
State 25 REMAINING t1:[1..1]
    2. [case [[42]]:] AND(t5=DagIndexerEvaluation(t0[0]), Not (DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: t5=DagIndexerEvaluation(t0[0])
    TrueBranch: 26
State 26 REMAINING t1:[1..1]
    2. [case [[42]]:] AND(Not (DagAssignmentEvaluation(t5=t2)), ?DagNonNullTest(t5), t6=DagPropertyEvaluation(t5.Length), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t5[0]), ?DagValueTest(t7 == ConstantValueI32(42: Int32)))
    Test: DagAssignmentEvaluation(t5=t2)
    TrueBranch: 27
*State 27 FAIL

```

</details>

Reply:
Never mind. I was thinking of a nested slice (`..[ ... ]`), but here we're testing a nested list.